erxupdate/ntopnghosts.sh

```diff
@@ -0,0 +1,429 @@
+    if [[ "$1" == "now" ]]; then
+	    date --utc +%s
+	else
+		date --utc --date "$1" +%s
+	fi
+}
+
+dateDiff (){ # source: https://www.unix.com/tips-and-tutorials/31944-simple-date-time-calulation-bash.html
+    case $1 in
+        -s)   sec=1;      shift;;
+        -m)   sec=60;     shift;;
+        -h)   sec=3600;   shift;;
+        -d)   sec=86400;  shift;;
+        *)    sec=86400;;
+    esac
+    dte1=$(date2stamp "$1")
+    dte2=$(date2stamp "$2")
+    diffSec=$((dte2-dte1))
+    if ((diffSec < 0)); then abs=-1; else abs=1; fi
+    echo $((diffSec/sec*abs))
+}
+
+sendmessage (){
+	RESULT=`curl -s "https://api.telegram.org/bot$TELEGRAM_API_KEY/sendMessage?chat_id=$TELEGRAM_CHAT_ID&text=$MESSAGE"`
+	#echo "RESULT=$RESULT"
+	OK_STR=`echo $RESULT |jq -r '."ok"'`
+	#echo "OK_STR=$OK_STR"
+	if [[ "$OK_STR" != "true" ]]; then
+		# Failed to update:  We're going to wait 30 seconds and then retry and then we'll give up.  This means that if we fail again that there 
+		# will be no notification about this host
+		sleep 31s
+		RESULT=`curl -s "https://api.telegram.org/bot$TELEGRAM_API_KEY/sendMessage?chat_id=$TELEGRAM_CHAT_ID&text=$MESSAGE"`
+		OK_STR=`echo $RESULT |jq -r '."ok"'`
+	fi
+	if [[ "$OK_STR" == "true" ]]; then
+		OK_STR="OK"
+	fi
+	echo "Sending telegram message about $TELEGRAMHOSTNAME result=$OK_STR"
+}
+
+#initialise variables - this section shouldn't be modified
+# HOSTLOG is used for temporarily storing the records of hosts we see in case we need to write them to the hostnameregister if TELEGRAMENABLE is set to YES
+HOSTLOG=
+CRLF="\n"
+HOSTNAME=
+MAC=
+
+#check we're not already running
+if [[ "`ps -ef |grep ntopnghosts.sh | wc -l`" -gt "4" ]]; then
+    # if our process is present already we don't want to run again so we'll stop (more than 4 processes with our name means we are already running)
+	exit
+fi
+
+# we need to check if the hostnamesubstitute file needs to be updated
+if [[ "$NAMESUBSTITUTEURL" != "" ]]; then
+	NAMESUBSTITUTESIZE=`curl -sI $NAMESUBSTITUTEURL --location --silent -H 'Accept-Encoding: gzip,deflate' |grep -i content-length |cut -d ' ' -f2`
+	NAMESUBSTITUTESIZE=(${NAMESUBSTITUTESIZE:0:-1})
+	#curl -sI $NAMESUBSTITUTEURL --location --silent --write-out 'size_download=%{size_download}\n'
+	if [[ -f "/tmp/hostnamesubstitute" ]]; then
+		FILESIZE=`wc -c /tmp/hostnamesubstitute |cut -d ' ' -f1`
+	else
+		FILESIZE=0
+		if [[ -f "/etc/nprobe/hostnamesubstitute" ]]; then
+			cp /etc/nprobe/hostnamesubstitute /tmp/hostnamesubstitute
+		fi
+	fi
+	#echo "Centrally managed Namesubstitute size: $NAMESUBSTITUTESIZE"
+	#echo "Local Namesubstitute size: $FILESIZE"
+	if [[ "$NAMESUBSTITUTESIZE" != "0" ]]; then	
+		if [[ "$FILESIZE" != "$NAMESUBSTITUTESIZE" ]]; then
+			echo "Downloading updated name substitute file"
+			curl -s $NAMESUBSTITUTEURL -o "/tmp/hostnamesubstitute_new"
+			FILESIZE=`wc -c /tmp/hostnamesubstitute_new|cut -d ' ' -f1`
+			if [[ "$FILESIZE" == "$NAMESUBSTITUTESIZE" ]]; then
```

Comment:
Use an hash function rather than the size to determine changes in a more reliable way.

Reply:
good idea but I don't want to bog down a router with cryptographic functions computing hashes of files, so instead I'll switch to using cmp to compare the file contents for any change directly.

Reply:
Additional note to the above response.  My comments were in relation to all the places in the script where I was using the filesize to determine if the file was the same or not.  However, they did not relate specifically to the instance above.  In the instance above we are checking the filesize on the remote server and comparing it with the local filesize so as to decide whether to download the remote file or not.  To my knowledge there is no way to compute a hash of a remote file without downloading it, and it was a design parameter not to download anything unnecessarily and therefore not to download it just to determine not to download it.  As a result, for determining whether to download the file or not we have to stick with using the filesize as that information IS available without downloading the file.

Reply:
The other time we use the filesize for comparison is immediately after downloading the file we confirm that we have downloaded a file of the size that we were anticipating.  This is to protect against the file download being interrupted and receiving an incomplete file and then accidentally using that incomplete file. Again, this is not suitable for replacing with a hash function as to do so we would have to know the original hash of the file on the remote server - which is not possible to my knowledge.